Постановка целей (CRUD)

dem9968 opened the pull request:
Постановка целей (CRUD)

dem9968 committed:
CRUD

MiraGlaceon commented on src/main/java/school/faang/user_service/config/MapperConfig.java:
```diff
@@ -0,0 +1,15 @@
+public class MapperConfig {
+
+    @Bean
+    public GoalMapper goalMapper() {
```
Ты когда создаешь мапперы в MapStruct, там можно сразу в аннотации указать что данный класс - это бин. Ты его уже указал: componentModel = "spring"

dem9968 replied:
Ручное создание бина нужно т.к. интаче не работает автовайринг для юзерСервиса

MiraGlaceon commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,175 @@
+    //todo не сделаны эндпоинты: получения целей + фильтры; завершения цели
+
+    @PostMapping
+    public ResponseEntity<GoalResponse> createGoalHandler(@RequestBody JsonNode rawRequest) throws JsonProcessingException {
```
1) Ты можешь не работать с jackson напрямую, spring все сделает за тебя. Поэтому работу с JsonNode можно опустить. Создай сразу dto, которое поступает как RequestBody. У тебя уже есть такое dto - GoalCreateDto. И все поля автоматически будут пропаршены из json

2) Запись response(() -> createGoal(goalCreateRq), HttpStatus.CREATED); нормальная и хорошая, но уточню, что ты можешь сразу делать return createGoal(goalCreateRq) и spring автоматически обернет ответ в ResponseEntity

dem9968 replied:
исправил

MiraGlaceon commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,175 @@
+    }
+
+    @PatchMapping(path = "/{goalId}")
+    public ResponseEntity<GoalResponse> updateGoalHandler(@PathVariable Long goalId, @RequestBody JsonNode rawRequest) throws JsonProcessingException {
```
У тебя контроллер должен только принимать запросы от клиента и делегировать работу сервисам по работе с бизнес-логикой. Поэтому лучше выносить всю логику в сервисы, в том числе и валидацию. Так код будет читаться понятнее и каждый класс будет отвечать за собственную роль

dem9968 replied:
исправил

MiraGlaceon commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,175 @@
+        return violations;
+    }
+
+    @ExceptionHandler(BusinessException.class)
```
Эксепшен хэндлеры можно оставить в контроллере, но лучше вынести их в отдельный класс. Для этого предназначены эдвайсоры для контроллера - ControllerAdvice. Можно создать класс, аннотировать его как ControllerAdvice и выписать туда все ExceptionHandlers. Так у тебя контроллер останется чистым и будет содержать логику по делегации задач, а обработчики ошибок будут агрегированы в одном месте.

В обработчиках ошибок так же можно сразу возвращать значение, а спринг сам обернет его в ResponseEntity. Статус ответа можно указывать аннотацией

dem9968 replied:
вынес в ControllerAdvice

MiraGlaceon commented on src/main/java/school/faang/user_service/dto/goal/GoalRequest.java:
```diff
@@ -0,0 +1,6 @@
+
+import java.io.Serializable;
+
+public interface GoalRequest extends Serializable {
```
Не нужно создавать специально классы для сериализации. Спринг работает отлично и сам все оборачивает реквесты и респонсы как надо. Ты так создаешь длинную цепочку зависимостей наследник-родитель. Рекомендую переходить с работы с jackson на работу с чистыми DTO

dem9968 replied:
интерфейс GoalRequest нужен для того чтобы одни фильтры валидации применять для разных типов запросов
интерфейс GoalResponse нужен для того чтобы метод response мог возвращать ответы разного формата

MiraGlaceon commented on src/main/java/school/faang/user_service/dto/goal/GoalCreateDto.java:
```diff
@@ -0,0 +1,9 @@
+import lombok.Data;
+
+@Data
+public class GoalCreateDto implements GoalRequest {
```
Написал ниже почему не стоит использовать GoalRequest. Лишнее наследование. 
По поводу аннотации Data разговаривали на созвоне, можешь оставить, но рекомендую использовать выборочно Getter, Setter и другие по необходимости

dem9968 replied:
про интерфейсы ответил выше.

MiraGlaceon commented on src/main/java/school/faang/user_service/dto/goal/GoalCreateDto.java:
```diff
@@ -0,0 +1,9 @@
+@Data
+public class GoalCreateDto implements GoalRequest {
+    private Long userId;
+    private GoalDto goal;
```
Это DTO - объект для передачи данных между слоями. Поэтому можешь в дто класть любые необходимые тебе для работы поля. Тут я вижу, что ты мог добавить userId сразу в GoalDto, а не создавать еще один класс-обертку

dem9968 replied:
GoalDto отдельно используется для метода апдейта цели, поэтому так

MiraGlaceon commented on src/main/java/school/faang/user_service/dto/goal/GoalErrorResponseDto.java:
```diff
@@ -0,0 +1,54 @@
+
+    @Data
+    @AllArgsConstructor
+    public static class Error implements Serializable {
```
1) Очень много лишних конструкций связанных с сериализацией. Спринг сам все сделает, работа с jackson реализована в спринге из коробки и относится к middlewere, тебе не нужно беспокоиться об этом

2) Если тебе нужен класс ошибки, то лучше сделать не внутренний, а отдельный класс ошибки и обрабатывать ее в ExceptionHandler. Так у тебя ошибка не будет привязана конкретно к Goal, плюс будет описано все в одном месте. Представь, то ли я должен искать описание ошибки внутри ДТО, то ли я сразу открою ExceptionHandler и сразу найду там обработку ошибки и оттуда же найду путь к описанию ошибки.

dem9968 replied:
на созвоне договорились не переделывать то что связано с обработкой ошибок, т.к. потребуется по сути все переписать заново

MiraGlaceon commented on src/main/java/school/faang/user_service/dto/goal/GoalErrorResponseDto.java:
```diff
@@ -0,0 +1,54 @@
+        this.description = errorCode.getDescription();
+    }
+
+    public GoalErrorResponseDto(BusinessException error) {
```
Это удалить нужно из дто. Тебе нужен ДТО, для возврата клиенту данных. Значит - это либо успешный ответ (статус 200), либо ошибочный ответ (404, 500 или другие ошибочные коды). Успешный ответ возвращается из контроллера и возвращает валидный ДТО. Ошибочный ответ возвращается из ExceptionHandler. Если что-то в логике пошло не так, то ты бросаешь ошибку и ловишь и обрабатываешь ее из ExceptionHandler

dem9968 replied:
на созвоне договорились не переделывать то что связано с обработкой ошибок, т.к. потребуется по сути все переписать заново

MiraGlaceon commented on src/main/java/school/faang/user_service/dto/goal/GoalIdDto.java:
```diff
@@ -0,0 +1,4 @@
+package school.faang.user_service.dto.goal;
+
+public record GoalIdDto(Long id) implements GoalResponse {
```
Оцени что здесь происходит:
1) У тебя есть класс GoalResponse, который ты создал для сериализации. Это лишнее, потому что спринг все сделает а тебя
2) У тебя есть класс GoalIdDto для одного поля, который уже имеет тип класс Long

То есть ты обернул валидный класс в еще один класс и расширил еще одним пустым классом. Принимай из контроллера сразу класс Long, он самодостаточный для передачи айдишника

dem9968 replied:
мне нужен был ответ в формате
{
    "goalId": 100500
}

MiraGlaceon commented on src/main/java/school/faang/user_service/dto/goal/GoalErrorResponseDto.java:
```diff
@@ -0,0 +1,54 @@
+        private int code;
+        private String textCode;
+        private String description;
+        private ErrorField field;
```
Собственно этот класс ErrorField тоже не нужен

dem9968 replied:
на созвоне договорились не переделывать то что связано с обработкой ошибок, т.к. потребуется по сути все переписать заново

MiraGlaceon commented on src/main/java/school/faang/user_service/entity/Violation.java:
```diff
@@ -0,0 +1,24 @@
+
+@Data
+@AllArgsConstructor
+public class Violation implements Serializable {
```
Рекомендую не делать конструкции, а делать устоявшийся и простой способ обрабатывать ошибки, уже описал выше как это сделать, но повторю. 
У тебя есть успешный исход запроса  и ошибочный. Если успех, то мы возвращаем код 200 и валидный ДТО ответа. Если ошибка, то лучше создать класс ошибки и из кода выбросить эту ошибку, а  потом обработать и передать в ExceptionHandler. 

Пример:
Пытаешься загрузить цели из базы по Id, но по такому айди нет Goal в базе. Тогда создаешь класс ошибку:
```
class GoalNotFoundException extends RuntimeException 
```

и просто делаешь конструктор с нужным сообщением:
```
public GoalNotFoundException() {
   super("Goal not found by id");
} 
```

Дополнительно можно в конструктор передать айдишник, который поступал в запросе.

Далее в коде просто делаешь throw new GoalNotFoundException если не нашел в БД цели по айдишнику. 

И все, далее создаешь ExceptiongHandler на ошибку GoalNotFoundException и там делаешь return ошибочного ДТО, либо просто ошибочную строку, часто этого достаточно и лишних классов для ошибок делать не требуется

dem9968 replied:
преследовал цель сначала собрать все ошибки, а затем кучей вернуть в ответе сервиса.

на созвоне договорились не переделывать то что связано с обработкой ошибок, т.к. потребуется по сути все переписать заново

MiraGlaceon commented on src/main/java/school/faang/user_service/entity/WorkSchedule.java:
```diff
@@ -42,7 +42,7 @@ public class WorkSchedule {
     private String timezone; // часовой пояс в формате IANA (например, "Europe/Moscow")
 
-    @ManyToOne
+    @OneToOne
```
Напиши, пожалуйста, для чего эти изменения. WorkSchedule не относится вроде к твоей задаче

dem9968 replied:
просто исправил логическую ошибку в энтити

MiraGlaceon commented on src/main/java/school/faang/user_service/enums/ErrorCode.java:
```diff
@@ -0,0 +1,35 @@
+
+@RequiredArgsConstructor
+@Getter
+public enum ErrorCode {
```
1) Есть несколько вариантов для хранения текстов ошибок.
Самый популярный - это просто прямо из кода возвращать сообщение, например, throw new BlaBlaException('Превышено число максимально допустимых целей для пользователя'). 
Есть второй популярный, это когда используется локализация ошибки и чтобы клиент видел ошибку на собственном языке, там подход похожий throw new BlaBlaExcepton(errorMessages.getMessageByKey("maxGoalsForUser")), типа такого
Твой вариант имеет место быть, встречал его и сам применял, но только там, где используются стандартизированные ошибки, когда на каждый случай есть принятый регламентом компании или гостом ответ и тому подобное. 
Рекомендую тебе использовать первый по популярности подход, когда ты можешь просто сразу написать текст ошибки сразу. На нашем проекте этого будет достаточно

2) Когда пишешь текст ошибок, то вноси конкретику, чтобы было понятно по какой причине произошла ошибка. Ошибки такого типа: "Ошибка бизнес логики",  мало что о себе говорят. Бизнес-логика обширное понятие. Хороший пример - это "Ментор с указанным ID не существует", плюс еще указать ID в тексте ошибки

dem9968 replied:
"Ошибка бизнес логики" - это верхнеуровневая ошибка. В дополнение к ней уровнем ниже всегда есть более узкая ошибка.

на созвоне договорились не переделывать то что связано с обработкой ошибок, т.к. потребуется по сути все переписать заново

MiraGlaceon commented on src/main/java/school/faang/user_service/enums/GoalAction.java:
```diff
@@ -0,0 +1,21 @@
+@RequiredArgsConstructor
+@Getter
+public enum GoalAction {
+    CREATE_GOAL("createGoal", "Create Goal"),
```
Рекомендую избавиться от этой конструкции. Ты можешь использовать агрегатор ошибок валидации, особенно будет уместно, когда данных для валидации много. Например, когда у тебя много полей на форме регистрации или много данных по проверке реквизитов и отчетных данных юр лица в банке. 
В нашей ситуации у нас просто несколько проверок и, если что-то нарушено, то просто сразу бросаем ошибку. Поэтому группировать их и собирать не имеет большого смысла.

Дополнительно скажу: Всегда задумывайся будет ли полезно, если я добавлю такой-то то или такой-то класс? Будет ли это хорошо или это просто усложнит код? Как потом я или мои коллеги будут его отлаживать, сопровождать или тестировать? И другие критические опросы

Есть принцип в программировании, называется KISS - keep it simple, stupid. То есть не нужно усложнять вещи

dem9968 replied:
удалил действия за ненадобностью после изменения апдейта цели с PATCH на PUT

MiraGlaceon commented on src/main/java/school/faang/user_service/exception/BusinessException.java:
```diff
@@ -0,0 +1,27 @@
+import java.util.List;
+
+@Getter
+public class BusinessException extends RuntimeException {
```
1) BusinessException - слишком абстрактное название. Нужно указывать конкретику в имени класса, чтобы тот кто читал код сразу догадывался о чем речь

2) Уже писал выше, что можно создавать ошибки таким путем, но поля  HttpStatus, ErrorCode и violations лишние. Ты проблемное место сразу осветишь в тексте ошибки, код и статус ответа клиент получит самостоятельно из http заголовка, эту информацию не нужно лишний раз прописывать в ответе

dem9968 replied:
на созвоне договорились не переделывать то что связано с обработкой ошибок, т.к. потребуется по сути все переписать заново

MiraGlaceon commented on src/main/java/school/faang/user_service/mapper/GoalMapper.java:
```diff
@@ -0,0 +1,34 @@
+@Mapper(componentModel = "spring", uses = {SkillServiceImpl.class})
+public abstract class GoalMapper {
+    @Autowired
+    protected UserServiceImpl userService;
```
Нет, тут грубая ошибка. Маппер - это маппер, который выполняет свои задачи по конвертации одного объекта в другой. Использовать сторонние сервисы, например, userService и подгружать что-то из базы не нужно. Вполне нормально использовать другие мапперы внутри других мапперов, к примеру

dem9968 replied:
проговорили на созвоне, решили что норм

MiraGlaceon commented on src/main/java/school/faang/user_service/repository/goal/GoalRepository.java:
```diff
@@ -23,6 +25,12 @@ INSERT INTO goal (title, description, parent_goal_id, status, created_at, update
             """)
     Goal create(String title, String description, Long parent);
 
+    @Query(nativeQuery = true, value = """
```
Когда что-то записываешь или обновляешь в базе через репозитории с использованием аннотации Query, то так же используй аннотацию Modifying. Для чтения данных из БД эта аннотация не нужна. Modifying указывает спрингу, что будут изменяться данные. Без этой аннотации спринг воспримет query как select-запрос для чтения данных .

Так же при изменении данных в БД рекомендую оборачивать в транзакцию

dem9968 replied:
Modifying работает только с методами которые возвращают void или int. У меня метод возвращает Goal

MiraGlaceon commented on src/main/java/school/faang/user_service/service/SkillServiceImpl.java:
```diff
@@ -0,0 +1,22 @@
+    private final SkillRepository skillRepository;
+
+    @Override
+    @Named("mapSkillIdsToSkillList")
```
Расскажи, пожалуйста, для чего ты использовал Named("mapSkillIdsToSkillList")

dem9968 replied:
для того чтобы мэпстракт смапил скилы с использованием указанного метода в классе SkillService

MiraGlaceon commented on src/main/java/school/faang/user_service/service/UserService.java:
```diff
@@ -0,0 +1,9 @@
+
+import java.util.Optional;
+
+public interface UserService {
```
Есть у меня подозрения, что SkillService и UserService уже делались до тебя. Обязательно перепроверь, пожалуйста

dem9968 replied:
я это и сделал по аналогии того как было сделано и залито в репозиторий

MiraGlaceon commented on src/main/java/school/faang/user_service/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,19 @@
+
+    @Override
+    public Optional<User> getUserEntityById(Long userId) {
+        return userRepository.findById(userId);
```
UserServiceImpl - это у тебя сервис с логикой, ты прям здесь можешь проверить что optional не пустой и обработать

dem9968 replied:
поменял

MiraGlaceon commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,225 @@
+import java.util.List;
+
+@Service
+@RequiredArgsConstructor(onConstructor_ = @Autowired)
```
Расскажи, пожалуйста, для чего ты используешь onConstructor_ = @Autowired

dem9968 replied:
не удалил лишнее после попыток разобраться с инжекцией в маппер без создания бина вручную.

MiraGlaceon commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,225 @@
+    }
+
+    @Override
+    @Transactional(propagation = Propagation.REQUIRED)
```
Расскажи для чего ты используешь Propagation для транзакции? Какое значение по умолчанию и почему ты остановился на REQUIRED типе?

dem9968 replied:
не удалил лишнее после попыток разобраться с транзакцией при рекурсивном вызове.

MiraGlaceon commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,225 @@
+    @Transactional(propagation = Propagation.REQUIRED)
+    public GoalIdDto createGoal(GoalCreateDto goalCreateRq) throws BusinessException {
+
+        return util.executeInTransaction(status -> {
```
1) Не понятно из кода что такое Util. Если делать какой-то класс, то лучше ему назначать понятное имя, чтобы было легко понять что он делает

2) Ты уже код выполняешь в транзакции, не нужно еще раз оборачивать в транзакцию

dem9968 replied:
переименовал в Helper.
ручное управление транзакциями нужно т.к. метод рекурсивно вызывает себя же, а спринг выполняет такие вызовы в отдельной транзакции.

MiraGlaceon commented on src/main/java/school/faang/user_service/service/goal/GoalServiceImpl.java:
```diff
@@ -0,0 +1,225 @@
+                );
+            }
+
+            //todo здесь по плану должны возвращаться данные созданной цели, но для упрощения возвращается только Id
```
Ты так оставил для упрощения как есть или оставил доработку на потом? Потому что если это просто комментарий, то убери todo, чтобы не путать

dem9968 replied:
удалил туду

MiraGlaceon commented on src/main/java/school/faang/user_service/util/Util.java:
```diff
@@ -0,0 +1,52 @@
+
+@Component
+@RequiredArgsConstructor
+public class Util {
```
Излишняя конструкция, спринг уже дает возможность оборачивать код в транзакцию через аннотацию. Использовать transactionTemplate можно в крайних частных случаях, у меня не получается придумать пример, где его можно использовать без аннотации, думаю в 95% случаев подойдет аннотация. На нашем проекте все можно сделать через аннотацию. Рекомендую отойти от использования транзакций вручную

dem9968 replied:
про ручное управление транзакцией написал в одном из комментов выше

MiraGlaceon commented on src/main/java/school/faang/user_service/validator/Validator.java:
```diff
@@ -0,0 +1,9 @@
+package school.faang.user_service.validator;
+
+public interface Validator<T, T2> {
+    ValidationResult validate(T object, T2 object2);
```
Уже писал по поводу валидации, в нашем случае достаточно сразу в коде проверить и бросить ошибку в случае нарушений

dem9968 replied:
на созвоне договорились не переделывать то что связано с обработкой ошибок, т.к. потребуется по сути все переписать заново

MiraGlaceon commented on src/main/java/school/faang/user_service/validator/goal/GoalDeadlineValidator.java:
```diff
@@ -0,0 +1,25 @@
+public class GoalDeadlineValidator implements Validator<Goal, GoalValidationParams> {
+
+    @Override
+    public ValidationResult validate(Goal goal, GoalValidationParams validationParams) {
```
То же самое, рекомендую убрать это и далее что связано с валидацией

dem9968 replied:
на созвоне договорились не переделывать то что связано с обработкой ошибок, т.к. потребуется по сути все переписать заново

dem9968 committed:
Валидации вынесены в отдельные классы
Вся логика убрана из контроллера
Апдейт цели теперь PUT и предполагает наличие всех полей в запросе
Эксепшн хэндлеры вынесены в ControllerAdvice

dem9968 committed:
Удалил лишние аннотации

dem9968 committed:
Убрал Optional в методе getUserEntityById
